examples/vision/handwriting_recognition.py

```diff
@@ -0,0 +1,508 @@
+            [0, 0],
+        ],
+    )
+
+    image = tf.transpose(image, perm=[1, 0, 2])
+    image = tf.image.flip_left_right(image)
+    return image
+
+
+"""
+If we just go with the plain resizing then the images would look like so:
+
+![](https://i.imgur.com/eqq3s4N.png)
+
+Notice how this resizing would have introduced unnecessary stretching.
+"""
+
+"""
+### Putting the utilities together
+"""
+
+batch_size = 64
+padding_token = 99
+image_width = 128
+image_height = 32
+
+
+def preprocess_image(image_path, img_size=(image_width, image_height)):
+    image = tf.io.read_file(image_path)
+    image = tf.image.decode_png(image, 1)
+    image = distortion_free_resize(image, img_size)
+    image = tf.cast(image, tf.float32) / 255.0
+    return image
+
+
+def vectorize_label(label):
+    label = char_to_num(tf.strings.unicode_split(label, input_encoding="UTF-8"))
+    length = tf.shape(label)[0]
+    pad_amount = max_len - length
+    label = tf.pad(label, paddings=[[0, pad_amount]], constant_values=padding_token)
+    return label
+
+
+def process_images_labels(image_path, label):
+    image = preprocess_image(image_path)
+    label = vectorize_label(label)
+    return {"image": image, "label": label}
+
+
+def prepare_dataset(image_paths, labels):
+    dataset = tf.data.Dataset.from_tensor_slices((image_paths, labels)).map(
+        process_images_labels, num_parallel_calls=AUTOTUNE
+    )
+    return dataset.batch(batch_size).cache().prefetch(AUTOTUNE)
+
+
+"""
+## Prepare `tf.data.Dataset` objects
+"""
+
+train_ds = prepare_dataset(train_img_paths, train_labels_cleaned)
+validation_ds = prepare_dataset(validation_img_paths, validation_labels_cleaned)
+test_ds = prepare_dataset(test_img_paths, test_labels_cleaned)
+
+"""
+## Visualize a few samples
+"""
+
+for data in train_ds.take(1):
+    images, labels = data["image"], data["label"]
+
+    _, ax = plt.subplots(4, 4, figsize=(15, 8))
+
+    for i in range(16):
+        img = images[i]
+        img = tf.image.flip_left_right(img)
+        img = tf.transpose(img, perm=[1, 0, 2])
+        img = (img * 255.0).numpy().clip(0, 255).astype(np.uint8)
+        img = img[:, :, 0]
+
+        # Gather indices where label!= padding_token.
+        label = labels[i]
+        indices = tf.gather(label, tf.where(tf.math.not_equal(label, padding_token)))
+        # Convert to string.
+        label = tf.strings.reduce_join(num_to_char(indices))
+        label = label.numpy().decode("utf-8")
+
+        ax[i // 4, i % 4].imshow(img, cmap="gray")
+        ax[i // 4, i % 4].set_title(label)
+        ax[i // 4, i % 4].axis("off")
+
+
+plt.show()
+
+"""
+We can notice that the content of original image is kept as intact as possible and has
+been accordingly padded. 
+"""
+
+"""
+## Model
+
+Our model will use the CTC loss as an endpoint layer. For a detailed understanding of the
+CTC loss, refer to [this post](https://distill.pub/2017/ctc/). 
+"""
+
+
+class CTCLayer(keras.layers.Layer):
```

Comment:
Do we have to make the loss a `Layer` instead of a loss function or `Loss` subclass?

Reply:
Our idea with this example is to keep it as close to the Captcha OCR example as possible while showing the bits that need to be changed. IMO, that helps to ensure a good reading experience. Ccing @AakashKumarNain if he has other points of view.

Reply:
When defining a normal loss function, we generally assume that the shape of `y_true` and `y_pred` are same which isn't the case here. Defining it as en endpoint layer makes it easy to calculate the length of inputs and the labels on the fly during training, and then pass the same to the `ctc_batch_cost(...)`